pkg/operator/bootstrapteardown/bootstrap_teardown_controller.go

```diff
@@ -0,0 +1,306 @@
+	}
+
+	if !etcdReady {
+		klog.Infof("Still waiting for the cluster-etcd-operator to bootstrap...")
+		return nil
+	}
+
+	kubeAPIServer, err := c.kubeAPIServerLister.Get("cluster")
+	if err != nil {
+		return err
+	}
+
+	kasReady := c.isKASReady(kubeAPIServer, c.configMapLister)
+	if !kasReady {
+		klog.Infof("Still waiting for the kube-apiserver to be ready...")
+		return nil
+	}
+
+	etcdEndpointExists := c.clusterMemberShipController.IsMember("etcd-bootstrap")
+	etcdMemberExists := c.clusterMemberShipController.IsEtcdMember("etcd-bootstrap")
+
+	if !etcdEndpointExists && !etcdMemberExists {
+		// set bootstrap removed condition
+		_, _, updateErr := v1helpers.UpdateStatus(c.operatorConfigClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:    clustermembercontroller.ConditionBootstrapRemoved,
+			Status:  operatorv1.ConditionTrue,
+			Reason:  "BootstrapNodeRemoved",
+			Message: "Etcd operator has scaled",
+		}))
+		if updateErr != nil {
+			c.eventRecorder.Warning("BootstrapTeardownErrorUpdatingStatus", updateErr.Error())
+			return updateErr
+		}
+		// return because no work left to do
+		return nil
+	} else {
+		c.eventRecorder.Eventf("ErrorRemovingBootstrapEndpoint", "Error removing bootstrap address from endpoint: %#v", err)
+		// explicitly setting condition
+		_, _, _ = v1helpers.UpdateStatus(c.operatorConfigClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:    clustermembercontroller.ConditionBootstrapRemoved,
+			Status:  operatorv1.ConditionFalse,
+			Reason:  "BootstrapNodeNotRemoved",
+			Message: fmt.Sprintf("Bootstrap node is not removed yet"),
+		}))
+	}
+
+	c.eventRecorder.Event("BootstrapTeardownController", "safe to remove bootstrap")
+
+	if err := c.clusterMemberShipController.RemoveBootstrapFromEndpoint(); err != nil {
+		return err
+	}
+
+	if err := c.clusterMemberShipController.EtcdMemberRemove("etcd-bootstrap"); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *BootstrapTeardownController) isKASReady(kasOperator *operatorv1.KubeAPIServer, configMapLister corev1listers.ConfigMapLister) bool {
+	revisionMap := map[int32]struct{}{}
+	uniqueRevisions := []int32{}
+
+	for _, nodeStatus := range kasOperator.Status.NodeStatuses {
+		revision := nodeStatus.CurrentRevision
+		if _, ok := revisionMap[revision]; !ok {
+			revisionMap[revision] = struct{}{}
+			uniqueRevisions = append(uniqueRevisions, revision)
+		}
+	}
+
+	// For each revision, check that the configmap for that revision contains the
+	// appropriate storageConfig
+	for _, revision := range uniqueRevisions {
+		configMapNameWithRevision := fmt.Sprintf("%s-%d", configMapName, revision)
+		configMap, err := configMapLister.ConfigMaps("openshift-kube-apiserver").Get(configMapNameWithRevision)
+		if err != nil {
+			klog.Errorf("doneApiServer: error getting configmap: %#v", err)
+			return false
+		}
+		if c.configMapHasRequiredValues(configMap) {
+			// if any 1 kube-apiserver pod has more than 1
```

Comment:
no, you need them all to have more than one.  If you only wait for one, the other kube-apiservers will suddenly start returning errors and we'll get bugs opened.

Reply:
we had it check for all kube-apiservers but we saw frequent time outs in CI waiting for correct config. Keeping it like this as long as we dont need to change.

Reply:
This is risky, but I'm willing to flap during install if I must. This is likely to break upgrades and I think we will have to revisit it.